### ISSUE TITLE
Add possibility to pass options to the HTTP client

### ISSUE THREAD

**Issue by philss** [matthewoden/libvault#6]
I'm trying to replace`vaultex` with `libvault`, but one problem is holding me back: I can't pass options to the HTTP client and therefore I cannot configure SSL using the `{:ssl, [cacertfile: path_to_file]}` option from `hackney`.

I know that I can use the global application to configure that, using `config :tesla, adapter: {Tesla.Adapter.Hackney, ssl: [certfile: "certs/client.crt"]}`, but this is not a good approach, since the certificate may be useful only for one host, and if I'm using Tesla for more than communicating with Vault, this can be a problem.

My proposal is to add the ability to pass arbitrary options to the HTTP client at runtime. I think the usage would be something like this:

```elixir
vault =
  Vault.new([
    engine: Vault.Engine.KVV2,
    http: Vault.HTTP.Tesla,
    http_client_opts: {Tesla.Adapter.Hackney, ssl: [certfile: "certs/client.crt"]},
    credentials: %{username: "username", password: "password"}
  ])
```

And the implementation of `Vault.HTTP` would have to pass down this new Keyword value to the HTTP adapter.

What do you think? Is this a valid issue? I can work on this if you agree. Thank you.

- Tesla config reference: https://github.com/teamon/tesla/blob/afca9522301a50c2740ffdcb5f00182b638fa548/lib/tesla.ex#L439

---

PS: This is a very nice and flexible library to communicate with Vault!
Please consider adding it to the list of Elixir libs: https://github.com/hashicorp/vault/blob/master/website/source/api/libraries.html.md#elixir

**Comment by matthewoden** [comment]
Absolutely a valid case! 

A few thoughts:
- what about `:http_options`, since the adapter key is just `:http`?
- For the actual options, I see what you did with mirroring the tesla client's configuration tuple - but I think for client flexibility, we should stick with a keyword list.

If you want to work on it, that'd be great! I'm excited to see what you put together.

And I'm glad you like the library! I considered adding it to the list, but I figured I'd wait until it had some use first.

**Comment by philss** [comment]
OK, `:http_options` looks better to me too. I think it's fine to stick with keyword lists, but I don't know how to do if the client does not specify the adapter for Tesla. I'm going to figure out 😄 .

**Comment by victorolinasc** [comment]
Just to add to the thread, one way to achieve this is to simply write a new HTTP adapter. Here is an example:

```elixir
defmodule CustomHttpAdapter do
  @moduledoc """
  Custom Tesla adapter for libvault
  """

  alias Tesla.Middleware.FollowRedirects

  @behaviour Vault.HTTP.Adapter

  @impl true
  def request(method, url, params, headers) do
    params = if params == "{}" and method in [:get, :delete, :head], do: nil, else: params
    Tesla.request(client(), method: method, url: url, headers: headers, body: params)
  end

  defp client do
    Tesla.client([{FollowRedirects, []}], adapter())
  end

  defp adapter do
    opts = # build your opts
    {Tesla.Adapter.Hackney, opts}
  end
end

# Then
Vault.new([
    # Other opts
    http: CustomHttpAdapter
  ])
```

I was looking into this issue cause I've migrated from vaultex too and this is how I've solved it currently.

P.S.: @matthewoden you are using two Tesla APIs: the `use Tesla` and the `Tesla.client()` directly. If you used only the first one we could configure specific adapter options on mix.exs with:

```elixir
config :tesla, Vault.HTTP.Tesla, adapter: # adapter options here
```

Using the second API it would need some custom options passing.

Hope this helps!

**Comment by philss** [comment]
@victorolinasc Good point, thanks! 

I think it's a good idea to have a way to pass options for the adapters, so I'm going to stick with the initial ideal (there is a PR comming).

About the config, I want to avoid setting the configuration for the entire application (in this case, for the "global" Tesla). But this is the way I got it working today:

```
Application.put_env(
  :tesla,
  :adapter,
  {Tesla.Adapter.Hackney, ssl_options: [cacertfile: System.get_env("VAULT_CA_FILE")]}
)
```

Which is kind of the same thing you suggested. So I assume that this already work.

**Comment by victorolinasc** [comment]
@philss they way you mention is the "global" tesla approach and you are right.

If, though, it used the pre 1.2 method and called things like: `request(....)` instead of `Tesla.request(...)` you could configure the adapter for that specific module with:

`config :tesla, SomeModule, adapter: # ...`. 

Anyway, your PR will be most welcome and will solve this the "proper" ® way :)

**Comment by philss** [comment]
The PR is open. I couldn't find a way to test that the options are being passed to the Tesla adapter. Please let me know if you have any ideas. Thanks!

**Comment by matthewoden** [comment]
Published as 0.2.0 - Again, thanks so much for this!

**Comment by philss** [comment]
@matthewoden thank you!
